osu.Server.QueueProcessor/BeatmapStatusWatcher.cs

```diff
@@ -0,0 +1,139 @@
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dapper;
+using MySqlConnector;
+
+namespace osu.Server.QueueProcessor
+{
+    /// <summary>
+    /// Provides insight into whenever a beatmap has changed status based on a user or system update.
+    /// </summary>
+    public static class BeatmapStatusWatcher
+    {
+        /// <summary>
+        /// Start a background task which will poll for beatmap sets with updates.
+        /// </summary>
+        /// <remarks>
+        /// The general flow of usage should be:
+        ///
+        /// // before doing anything else, start polling.
+        /// // it's important to await the completion of this operation to ensure no updates are missed.
+        /// using var pollingOperation = await StartPollingAsync(updates, callback);
+        ///
+        /// void callback(BeatmapUpdates u)
+        /// {
+        ///     foreach (int beatmapSetId in u.BeatmapSetIDs)
+        ///     {
+        ///         // invalidate anything related to `beatmapSetId`
+        ///     }
+        /// }
+        /// </remarks>
+        /// <param name="callback">A callback to receive information about any updated beatmap sets.</param>
+        /// <param name="pollMilliseconds">The number of milliseconds to wait between polls. Starts counting from response of previous poll.</param>
+        /// <param name="limit">The maximum number of beatmap sets to return in a single response.</param>
+        /// <returns>An <see cref="IDisposable"/> that should be disposed of to stop polling.</returns>
+        public static async Task<IDisposable> StartPollingAsync(Action<BeatmapUpdates> callback, int pollMilliseconds = 10000, int limit = 50)
+        {
+            var initialUpdates = await GetUpdatedBeatmapSetsAsync(limit: limit);
+            return new PollingBeatmapStatusWatcher(initialUpdates.LastProcessedQueueID, callback, pollMilliseconds, limit);
+        }
+
+        /// <summary>
+        /// Check for any beatmap sets with updates since the provided queue ID.
+        /// Should be called on a regular basis. See <see cref="StartPollingAsync"/> for automatic polling after the first call.
+        /// </summary>
+        /// <param name="lastQueueId">The last checked queue ID, ie <see cref="BeatmapUpdates.LastProcessedQueueID"/>.</param>
+        /// <param name="limit">The maximum number of beatmap sets to return in a single response.</param>
+        /// <returns>A response containing information about any updated beatmap sets.</returns>
+        public static async Task<BeatmapUpdates> GetUpdatedBeatmapSetsAsync(int? lastQueueId = null, int limit = 50)
+        {
+            using MySqlConnection connection = await DatabaseAccess.GetConnectionAsync();
+
+            if (lastQueueId.HasValue)
+            {
+                var items = (await connection.QueryAsync<bss_process_queue_item>("SELECT * FROM bss_process_queue WHERE queue_id > @lastQueueId LIMIT @limit", new
```

Comment:
Not sure how much it matters but here's a loose thought: the way this is written sort of seems to implicitly rely on this query being `ORDER BY queue_id ASC`. Which it is *probably* going to be, but I'm not sure anything in the SQL standard actually *guarantees* that a `SELECT` will respect primary key ordering, and reading stuff like https://dev.mysql.com/doc/refman/8.4/en/limit-optimization.html having `LIMIT` in can apparently alter row ordering in some cases.

Maybe worth specifying explicitly, maybe not. Dunno.

Reply:
Let's explicitly enforce it. I'll make the change.